src/iop/ashift.c

```diff
@@ -0,0 +1,938 @@
+                                                                                                             \
+    const c_type epsilon = 1e-7f;                                                                            \
+    if(fabs(det) < epsilon) return 1;                                                                        \
+                                                                                                             \
+    const c_type invDet = 1.0 / det;                                                                         \
+                                                                                                             \
+    B(1, 1) = invDet * (A(3, 3) * A(2, 2) - A(3, 2) * A(2, 3));                                              \
+    B(1, 2) = -invDet * (A(3, 3) * A(1, 2) - A(3, 2) * A(1, 3));                                             \
+    B(1, 3) = invDet * (A(2, 3) * A(1, 2) - A(2, 2) * A(1, 3));                                              \
+                                                                                                             \
+    B(2, 1) = -invDet * (A(3, 3) * A(2, 1) - A(3, 1) * A(2, 3));                                             \
+    B(2, 2) = invDet * (A(3, 3) * A(1, 1) - A(3, 1) * A(1, 3));                                              \
+    B(2, 3) = -invDet * (A(2, 3) * A(1, 1) - A(2, 1) * A(1, 3));                                             \
+                                                                                                             \
+    B(3, 1) = invDet * (A(3, 2) * A(2, 1) - A(3, 1) * A(2, 2));                                              \
+    B(3, 2) = -invDet * (A(3, 2) * A(1, 1) - A(3, 1) * A(1, 2));                                             \
+    B(3, 3) = invDet * (A(2, 2) * A(1, 1) - A(2, 1) * A(1, 2));                                              \
+    return 0;                                                                                                \
+  }
+
+#define A(y, x) src[(y - 1) * 3 + (x - 1)]
+#define B(y, x) dst[(y - 1) * 3 + (x - 1)]
+/** inverts the given 3x3 matrix */
+generate_mat3inv_body(float, A, B)
+
+int mat3inv(float *const dst, const float *const src)
+{
+  return mat3inv_float(dst, src);
+}
+
+generate_mat3inv_body(double, A, B)
+#undef B
+#undef A
+#undef generate_mat3inv_body
+
+void mat3mulv(float *dst, const float *const mat, const float *const v)
+{
+  for(int k = 0; k < 3; k++)
+  {
+    float x = 0.0f;
+    for(int i = 0; i < 3; i++) x += mat[3 * k + i] * v[i];
+    dst[k] = x;
+  }
+}
+
+void mat3mul(float *dst, const float *const m1, const float *const m2)
+{
+  for(int k = 0; k < 3; k++)
+  {
+    for(int i = 0; i < 3; i++)
+    {
+      float x = 0.0f;
+      for(int j = 0; j < 3; j++) x += m1[3 * k + j] * m2[3 * j + i];
+      dst[3 * k + i] = x;
+    }
+  }
+}
+
+void _print_roi(const dt_iop_roi_t *roi, const char *label)
```

Comment:
mat3mul() etc. should go static and _print_roi() should finally go away when development is finished.